src/ring_theory/graded_ring/basic.lean

```diff
@@ -0,0 +1,67 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser, Kevin Buzzard, Jujian Zhang
+-/
+import algebra.direct_sum.algebra
+import data.mv_polynomial
+import ring_theory.polynomial.homogeneous
+
+/-! # Typeclass for graded ring
+For definition of a ring `R` being graded by `A : ι → add_subgroup R`, see doc string of
+`graded_ring`.
+
+- `graded_ring.decompose : R → ⨁ i, A i` and `graded_ring.recompose : ⨁ i, A i → R` are the ring
+isomorphism between `R` and `⨁ i, A i` if `R` is graded by `A`.
+- `graded_ring.complete_lattice.independent` states that the `A` is independent in the sense that
+for any `i : ι`, `A i` and `⨆ (j ≠ i), A j` intersect trivially. The most noticable consequence of
+this is that `A i` and `A j` intersects trivally for distinct `i` and `j`.
+- `graded_ring.proj R A i r` is the degree `i : ι` component of `r : R`.
+- `graded_ring.support R A r` is the `finset ι` containing the `i : ι` such that the degree `i`
+component of `r` is not zero.
+- `is_homogeneous R A x` states that `x ∈ A i` for some `i : ι`.
+- `homogeneous_element R A` is the subtype of `R` of all `x` such that `is_homogeneous R A x`.
+- `mv_polynomial_is_graded` provides an instance saying that `mv_polynomial R σ` is ring graded by
+its homogeneous components.
+-/
+
+open_locale direct_sum big_operators
+
+section graded_ring
+variables (R : Type*) [ring R] {ι : Type*} (A : ι → add_subgroup R)
+  [decidable_eq ι] [add_comm_monoid ι]
+
+/-- A graded ring is a `ring R` such that `R` can be decomposed into a collection of
+  `add_subgroups R` indexed by `ι` such that the connonical map `R → ⨁ i, A i` is a bijective map
+  respecting multiplication, i.e. product of an element of degree `i` and an element of degree `j`
+  is an element of degree `i + j`.
+-/
+class graded_ring :=
+( decompose : R → ⨁ i, A i)
+( left_inv : function.left_inverse decompose (direct_sum.add_subgroup_coe A) )
+( right_inv : function.right_inverse decompose (direct_sum.add_subgroup_coe A) )
+( one_degree_zero : (1 : R) ∈ A 0 )
+( mul_respect_grading : ∀ {i j : ι} {a b : R}, a ∈ A i → b ∈ A j → a * b ∈ A (i + j))
```

Comment:
My hunch is that this might be easier to work with as
```suggestion
class graded_ring.core :=
( decompose : R → ⨁ i, A i)
( one_degree_zero : (1 : R) ∈ A 0 )
( mul_respect_grading : ∀ {i j : ι} {a b : R}, a ∈ A i → b ∈ A j → a * b ∈ A (i + j))

class graded_ring extends graded_ring_core R A :=
( left_inv : function.left_inverse decompose (direct_sum.add_subgroup_coe A) )
( right_inv : function.right_inverse decompose (direct_sum.add_subgroup_coe A) )
```

As that gives you enough to prove the `direct_sum.gsemiring` instance and the `(⨁ i, A i) →+* R` morphism early, which you can then use to help with the proofs of `left_inv` and `right_inv`